crates/turborepo-auth/src/lib.rs

```diff
@@ -12,19 +12,139 @@ mod ui;
+        client
+            .get_metadata(self.into_inner())
+            .await
+            .map_err(Error::from)
+    }
+    /// Returns the underlying token string.
+    pub fn into_inner(&self) -> &str {
         match self {
-            Self::Existing(s) => s,
-            Self::New(s) => s,
+            Self::Existing(token) | Self::New(token) => token.as_str(),
+        }
+    }
+}
+
+fn current_unix_time() -> u128 {
+    use std::time::{SystemTime, UNIX_EPOCH};
+    SystemTime::now()
+        .duration_since(UNIX_EPOCH)
+        .expect("Time went backwards")
+        .as_millis()
+}
+
+// As of the time of writing, this should always be true, since a token that
+// isn't active returns an error when fetching metadata for the token.
+fn is_token_active(metadata: &ResponseTokenMetadata) -> bool {
+    let current_time = current_unix_time();
+    let active_at = metadata.active_at;
+
+    let active_less_than_current = active_at <= current_time;
+    let all_scopes_active = metadata.scopes.iter().all(|scope| {
+        scope
+            .expires_at
+            .map_or(true, |expires_at| expires_at > current_time)
+    });
+
+    // Not all scopes have an expiration date, so we need to check if all of them
+    // are expired. If there isn't an expiration date, we assume they are infinite
+    // and therefore cannot be expired.
+    active_less_than_current && all_scopes_active
+}
+
+#[cfg(test)]
+mod tests {
+    use turborepo_vercel_api::token::Scope;
+
+    use super::*;
+
+    #[test]
+    fn test_is_token_active() {
+        let current_time = current_unix_time();
+        let quick_scope = |expiry| Scope {
+            expires_at: expiry,
+            scope_type: "".to_string(),
+            origin: "".to_string(),
+            created_at: 0,
+            team_id: None,
+        };
+        let mock_response = |active_at, scopes| ResponseTokenMetadata {
+            active_at,
+            scopes,
+            // These fields don't matter in the test
+            id: "".to_string(),
+            name: "".to_string(),
+            token_type: "".to_string(),
+            origin: "".to_string(),
+            created_at: 0,
+        };
+
+        let cases = vec![
+            // Case: Token active, no scopes (implicitly infinite)
+            (current_time - 10, vec![], true),
+            // Case: Token active, one scope without expiration
+            (current_time - 10, vec![quick_scope(None)], true),
+            // Case: Token active, one scope expired
+            (
+                current_time - 10,
+                vec![quick_scope(Some(current_time - 1))],
+                false,
+            ),
+            // Case: Token active, one scope not expired
+            (
+                current_time - 10,
+                vec![quick_scope(Some(current_time + 11))],
+                true,
+            ),
+            // Case: Token inactive (future `active_at`)
+            (
+                current_time + 10,
```

Comment:
I realize that this is unlikely, but I think we should proactively bump this threshold since there's a chance this could fail if it takes 10ms or more to perform the previous 4 test cases.

Also, I don't think this affects us, but the big "SystemTime is not monotonic" warning on the [docs](https://doc.rust-lang.org/std/time/struct.SystemTime.html#) makes a little concerned about some possible drift that causes these tests fail.

Reply:
I'll just do like, 1000 since this is a test that's supposed to be inactive.

The monotonic issue is valid, and I'm purely going off of vibes and instinct in saying that I don't think it'll be an issue. If it _does_ become an issue I can hard code something, but that feels a little more flakey / less exhaustive as a test